# **NAFISUL ISLAM**

## **Individual\_GroupNumber#1\_HW#3**

# 10 Custom Proposition
# Chat-GPT was used on some Proposition

# Proposition 1:

This query provides an aggregated view of **customer transactions** along with the corresponding **cities** where they are located. It calculates the **total number of transactions** and the **total transaction amount** for each customer, grouped by city.

The query filters the results to show only customers who have generated a **positive transaction amount**.

## Key Insights:

- The **total transaction activity** for each customer, represented by the number of transactions.
- The **city** in which the customer is located.
- The **revenue contribution** of each customer, showing only those who have positive transaction totals.

## Business Use Cases:

- Identifying **top-performing customers** and cities in terms of revenue.
- Analyzing **sales distribution** across various regions.
- Segmenting customers based on **transaction volume** and revenue contribution.
- Filtering out customers who do not generate revenue for the business.

The results are ordered by the **total transaction amount**, with the highest revenue-generating customers appearing first.

In [27]:
USE WideWorldImporters;
GO
SELECT 
    c.CustomerName, 
    city.CityName,
    COUNT(ct.CustomerTransactionID) AS TotalTransactions,
    SUM(ct.TransactionAmount) AS TotalAmount
FROM sales.customers AS c
JOIN sales.customerTransactions AS ct
    ON c.CustomerID = ct.CustomerID
JOIN Application.cities AS city
    ON c.DeliveryCityID = city.CityID
GROUP BY 
    c.CustomerName, 
    city.CityName
HAVING 
    SUM(ct.TransactionAmount) > 0
ORDER BY 
    TotalAmount DESC;



Commands completed successfully.

(39 rows affected)

Total execution time: 00:00:00.060

CustomerName CityName TotalTransactions TotalAmount Wingtip Toys (Head Office) Orchard Mesa 23233 97053.58 Tailspin Toys (Head Office) Lisco 23384 56435.84 Kumar Naicker Oconto Falls 216 13212.12 Linh Dao Mashulaville 253 6728.65 Rodrigo Figueiredo Blackville 199 5975.40 Shyam Poddar Tenstrike 220 5767.83 Dominic Davignon Twin Peaks 225 5732.98 Lilli Sokk Gresston 243 5689.31 Ana Florea Buchanan Lake Village 209 5511.15 Liidia Lepp Oakpark 218 5422.25 Satish Mittal Dry Run 233 5157.75 Isa Hulsegge Saxapahaw 155 4713.85 Suparna Bhattacharya Tiff City 187 4127.93 Narendra Tickoo Dillon 181 4038.80 Sabine Alksne Grannis 203 3876.65 Kertu Sokk Greig 204 3545.22 Cosmina Vlad Cañon 218 3427.58 Marcela Lucescu Ewen 206 3276.35 Emma Van Zant Wray 127 3091.20 Nils Kaulins Greycliff 185 3034.85 Geza Roman Oval 189 2685.25 Manca Hrastovsek Southside 195 2217.33 Meera Patel Hanaford 240 2157.40 Can ozcan Oley 197 1910.73 Neil Farrelly Sabinsville 202 1834.25 Taneli Riutta Sandy Level 198 1517.08 Alena Kellnerova Segundo 234 1456.40 Baalaamjali Devulapalli Puerto de Luna 224 1361.60 Lakshmi Benipal Lompoc 243 1361.60 Joel Carrillo Drumright 229 1223.60 Olya Izmaylov Flowella 222 745.20 Pavel Bogdanov New Braintree 229 613.18 Crina Grasu Tamora 208 449.65 Elina Kaleja Golden Meadow 257 426.08 Prabodh Nair Tschetter Colony 249 365.70 Luis Saucedo Linden 43 325.45 Milinka Zujovic Zurich 194 235.75 Sabine Zalite Haltom City 226 201.25 Tuulikki Linna Otterville 160 104.65

# Proposition 2:

This query calculates the **average transaction amount** for each customer and the city in which they are located. It helps identify customers who typically generate larger transaction values, providing insights into high-value clients and the cities they are associated with.

## Key Insights:

- The **average value of transactions** per customer, which can help assess the spending behavior of different customers.
- The **city** where each customer is located, allowing for geographic analysis of transaction patterns.
- Identification of customers who generate **above-average transaction amounts**, which may signify high-value clients or key business partners.

## Business Use Cases:

- **Customer Value Analysis**: Identifying high-value customers based on their average transaction amount, which can guide personalized marketing or loyalty programs.
- **Geographic Insights**: Understanding transaction trends in different cities, which can inform regional sales strategies.
- **Revenue Optimization**: Focusing business efforts on cities and customers that consistently generate larger transactions to optimize revenue streams.

The results are sorted in descending order of the **average transaction amount**, allowing businesses to easily identify the highest average transaction values.

In [28]:
USE WideWorldImporters;
GO

SELECT 
    c.CustomerName, 
    city.CityName, 
    AVG(ct.TransactionAmount) AS AverageTransactionAmount
FROM sales.customers AS c
JOIN sales.customerTransactions AS ct
    ON c.CustomerID = ct.CustomerID
JOIN Application.cities AS city
    ON c.DeliveryCityID = city.CityID
GROUP BY 
    c.CustomerName, 
    city.CityName
ORDER BY 
    AverageTransactionAmount DESC;


Commands completed successfully.

(263 rows affected)

Total execution time: 00:00:00.078

CustomerName CityName AverageTransactionAmount Kumar Naicker Oconto Falls 61.167222 Isa Hulsegge Saxapahaw 30.411935 Rodrigo Figueiredo Blackville 30.027135 Linh Dao Mashulaville 26.595454 Ana Florea Buchanan Lake Village 26.369138 Shyam Poddar Tenstrike 26.217409 Dominic Davignon Twin Peaks 25.479911 Liidia Lepp Oakpark 24.872706 Emma Van Zant Wray 24.340157 Lilli Sokk Gresston 23.412798 Narendra Tickoo Dillon 22.313812 Satish Mittal Dry Run 22.136266 Suparna Bhattacharya Tiff City 22.074491 Sabine Alksne Grannis 19.096798 Kertu Sokk Greig 17.378529 Nils Kaulins Greycliff 16.404594 Marcela Lucescu Ewen 15.904611 Cosmina Vlad Cañon 15.722844 Geza Roman Oval 14.207671 Manca Hrastovsek Southside 11.370923 Can ozcan Oley 9.699137 Neil Farrelly Sabinsville 9.080445 Meera Patel Hanaford 8.989166 Taneli Riutta Sandy Level 7.662020 Luis Saucedo Linden 7.568604 Alena Kellnerova Segundo 6.223931 Baalaamjali Devulapalli Puerto de Luna 6.078571 Lakshmi Benipal Lompoc 5.603292 Joel Carrillo Drumright 5.343231 Wingtip Toys (Head Office) Orchard Mesa 4.177401 Olya Izmaylov Flowella 3.356756 Pavel Bogdanov New Braintree 2.677641 Tailspin Toys (Head Office) Lisco 2.413438 Crina Grasu Tamora 2.161778 Elina Kaleja Golden Meadow 1.657898 Prabodh Nair Tschetter Colony 1.468674 Milinka Zujovic Zurich 1.215206 Sabine Zalite Haltom City 0.890486 Tuulikki Linna Otterville 0.654062 Surendra Sahu Old River-Winfree 0.000000 Akhil Nadar Schaal 0.000000 Nicolo Cattaneo Schoharie 0.000000 Dipti Shah Scioto Furnace 0.000000 Damodar Shenoy Scott 0.000000 Alinne Matos Sea Island 0.000000 Jackson Kolios Rail Road Flat 0.000000 Jayanta Thakur Raven 0.000000 Daniella Cavalcante Ravia 0.000000 Vedrana Kovacevic Reader 0.000000 Nhung Thach Red Mesa 0.000000 Serdar ozden Richland 0.000000 David Jaramillo Ridgeway 0.000000 Chaayaadaevi Sonti Roachtown 0.000000 Adriana Pena Roberta Mill 0.000000 Maryann Huddleston Robertsburg 0.000000 Emma Salpa Rogersville 0.000000 Taj Syme Oakshade 0.000000 Ganesh Majumdar Orland Hills 0.000000 Gopalgobinda Sikdar Orpha 0.000000 Cong Hoa Palatine Bridge 0.000000 Amrita Ganguly Palmas del Mar 0.000000 Madhu Dwivedi Panton 0.000000 Mahavir Sonkar Pastura 0.000000 Johanna Hoornstra Petronila 0.000000 Valentin Tirlea Pharr 0.000000 Staffan Persson Phelps City 0.000000 Lorin Astrom Port Alto 0.000000 Vladimir Henzl Port Chilkoot 0.000000 Damodara Trivedi Tarzan 0.000000 Ludmila Smidova Tempe 0.000000 Leila Carvalho Spillertown 0.000000 Daakshaayaani Sankaramanchi Spillville 0.000000 Fabrice Cloutier Staley 0.000000 Anindya Ghatak Switz City 0.000000 Ebru Erdogan Seiling 0.000000 Eugen Agafitei Shawboro 0.000000 Eric Torres Shell 0.000000 Gasper Havzija Shell Knob 0.000000 Gayatri Gajula Shenorock 0.000000 Olafs Rozitis Sherwin 0.000000 Jaroslav Fisar Sigel 0.000000 Shi Tu Sinclair 0.000000 Pari Hosseini Soham 0.000000 Isidora Morales Twodot 0.000000 Malorie Bousquet Valley View Park 0.000000 Libuse Kuchtova Van Alstyne 0.000000 Svetlana Todorovic Venersborg 0.000000 Aive Petrov Virgil 0.000000 Amarakumaar Gadiyaram Vollmar 0.000000 Philip Walker Waite Hill 0.000000 Yves Belisle Walapai 0.000000 Abhra Ganguly Walker Valley 0.000000 Gunnar Lohmus Wanaque 0.000000 Rajeev Sandhu Wasco 0.000000 Be Trang Watonga 0.000000 Jana Fialova Weigelstown 0.000000 Bhadram Kamasamudram Wells 0.000000 Drishti Bose Wesson 0.000000 Leyla Asef zade West Elkton 0.000000 Devraj Rao Westwater 0.000000 Hoc Tran White Horse Beach 0.000000 Kristiina Ivanov Wounded Knee 0.000000 Magdalena Michnova Wrangell 0.000000 Bala Dixit Termo 0.000000 Kamila Michnova The Colony 0.000000 Darshana Madhunapantula Tolchester Beach 0.000000 Sercan Celik Tolono 0.000000 David safranek Toughkenamon 0.000000 Tomo Vidovic Tremont 0.000000 Ingrida Zeltina Trilby 0.000000 Manjunatha Karnik Wright City 0.000000 Volkan senturk Wyndham 0.000000 Amet Shergill Yampa 0.000000 Kristi Part Yerington 0.000000 Bimla Prabhupāda Yermo 0.000000 Daevasree Samavedam Golf View 0.000000 Jai Lamble Gri

# Proposition 3:

This query retrieves **customer transactions** that represent payments received between **May 1, 2015, and May 31, 2015**. It provides insights into the specific types of transactions related to customer payments and allows for a deeper understanding of the inflow of payments within a defined time frame.

## Key Insights:

- The **transaction details** for payments received from customers during May 2015.
- The **transaction date**, amount, and type are retrieved to focus on payments received.
- The transactions are limited to those with the type **"Customer Payment Received"**, which helps filter out irrelevant transaction types such as refunds or invoices.

## Business Use Cases:

- **Cash Flow Analysis**: By focusing on customer payments within a specific time period, the business can analyze the inflow of funds.
- **Transaction Monitoring**: Helps finance and sales teams monitor payment behavior from customers, such as which customers made payments during the given time period.
- **Payment Auditing**: Provides a clear record of all payments received in a specified month, which can be used for accounting and reconciliation purposes.

The query filters transactions for **May 2015** and sorts the results by the **transaction date** in descending order, showing the most recent payments first.

In [30]:
USE WideWorldImporters;
GO
SELECT 
    ct.CustomerID, 
    ct.TransactionDate, 
    ct.TransactionAmount, 
    tt.TransactionTypeName
FROM sales.customerTransactions AS ct
JOIN Application.TransactionTypes AS tt
    ON ct.TransactionTypeID = tt.TransactionTypeID
WHERE tt.TransactionTypeName = 'Customer Payment Received'
AND ct.TransactionDate BETWEEN '2015-05-01' AND '2015-05-31'
ORDER BY ct.TransactionDate DESC;




Commands completed successfully.

(761 rows affected)

Total execution time: 00:00:00.021

CustomerID TransactionDate TransactionAmount TransactionTypeName 1 2015-05-31 -35003.01 Customer Payment Received 401 2015-05-31 -27858.81 Customer Payment Received 806 2015-05-31 -2083.80 Customer Payment Received 816 2015-05-31 -2577.15 Customer Payment Received 842 2015-05-31 -8373.15 Customer Payment Received 847 2015-05-31 -333.50 Customer Payment Received 853 2015-05-31 -296.70 Customer Payment Received 880 2015-05-31 -2990.00 Customer Payment Received 885 2015-05-31 -3426.08 Customer Payment Received 890 2015-05-31 -550.28 Customer Payment Received 903 2015-05-31 -2348.07 Customer Payment Received 911 2015-05-31 -2217.20 Customer Payment Received 913 2015-05-31 -1472.00 Customer Payment Received 918 2015-05-31 -155.25 Customer Payment Received 920 2015-05-31 -11186.05 Customer Payment Received 926 2015-05-31 -87.40 Customer Payment Received 941 2015-05-31 -4325.15 Customer Payment Received 947 2015-05-31 -876.30 Customer Payment Received 993 2015-05-31 -495.65 Customer Payment Received 1005 2015-05-31 -495.65 Customer Payment Received 1016 2015-05-31 -1242.00 Customer Payment Received 1024 2015-05-31 -932.65 Customer Payment Received 1026 2015-05-31 -270.25 Customer Payment Received 1037 2015-05-31 -1044.20 Customer Payment Received 1045 2015-05-31 -898.15 Customer Payment Received 1 2015-05-30 -100227.45 Customer Payment Received 401 2015-05-30 -73087.74 Customer Payment Received 802 2015-05-30 -1334.00 Customer Payment Received 820 2015-05-30 -3827.20 Customer Payment Received 834 2015-05-30 -524.40 Customer Payment Received 836 2015-05-30 -4062.95 Customer Payment Received 842 2015-05-30 -745.20 Customer Payment Received 844 2015-05-30 -2395.45 Customer Payment Received 846 2015-05-30 -1956.15 Customer Payment Received 863 2015-05-30 -587.42 Customer Payment Received 864 2015-05-30 -947.60 Customer Payment Received 865 2015-05-30 -2987.70 Customer Payment Received 867 2015-05-30 -704.95 Customer Payment Received 883 2015-05-30 -5989.20 Customer Payment Received 892 2015-05-30 -73.60 Customer Payment Received 895 2015-05-30 -1229.35 Customer Payment Received 896 2015-05-30 -5022.05 Customer Payment Received 897 2015-05-30 -1490.40 Customer Payment Received 903 2015-05-30 -2339.79 Customer Payment Received 906 2015-05-30 -876.76 Customer Payment Received 924 2015-05-30 -3174.00 Customer Payment Received 938 2015-05-30 -303.60 Customer Payment Received 939 2015-05-30 -1738.80 Customer Payment Received 947 2015-05-30 -5091.97 Customer Payment Received 952 2015-05-30 -9871.60 Customer Payment Received 959 2015-05-30 -285.20 Customer Payment Received 962 2015-05-30 -3697.25 Customer Payment Received 975 2015-05-30 -15494.01 Customer Payment Received 1001 2015-05-30 -4364.25 Customer Payment Received 1006 2015-05-30 -6590.65 Customer Payment Received 1007 2015-05-30 -4511.45 Customer Payment Received 1018 2015-05-30 -3891.60 Customer Payment Received 1019 2015-05-30 -7130.00 Customer Payment Received 1023 2015-05-30 -1069.50 Customer Payment Received 1029 2015-05-30 -186.30 Customer Payment Received 1031 2015-05-30 -3956.00 Customer Payment Received 1045 2015-05-30 -396.75 Customer Payment Received 1 2015-05-29 -45470.08 Customer Payment Received 401 2015-05-29 -102253.00 Customer Payment Received 805 2015-05-29 -2755.29 Customer Payment Received 809 2015-05-29 -598.86 Customer Payment Received 837 2015-05-29 -4306.75 Customer Payment Received 842 2015-05-29 -9269.00 Customer Payment Received 846 2015-05-29 -1888.30 Customer Payment Received 849 2015-05-29 -105.80 Customer Payment Received 859 2015-05-29 -1389.20 Customer Payment Received 868 2015-05-29 -644.00 Customer Payment Received 900 2015-05-29 -522.56 Customer Payment Received 912 2015-05-29 -1138.50 Customer Payment Received 916 2015-05-29 -612.95 Customer Payment Received 925 2015-05-29 -1025.80 Customer Payment Received 952 2015-05-29 -216.20 Customer Payment Received 961 2015-05-29 -333.50 Customer Payment Received 962 2015-05-29 -1518.23 Customer Payment Receiv

# Proposition 4:

This query provides an analysis of sales across **weekdays** and **years** to identify which days of the week have the most orders for each year. It calculates the total number of orders placed on each weekday, grouped by year, helping to identify trends in customer purchasing behavior over time.

## Key Insights:

- **Sales by Weekday**: The query shows which days of the week have the highest number of orders for each year.
- **Yearly Comparison**: It provides a breakdown of sales activity across different years, allowing businesses to compare how customer ordering patterns have changed over time.
- **Order Patterns**: This analysis helps to identify consistent trends, such as which days of the week consistently generate more sales.

## Business Use Cases:

- **Operational Planning**: By identifying the busiest days of the week for each year, businesses can optimize staffing, resources, and inventory levels for the most active sales days.
- **Year-over-Year Sales Analysis**: The query helps businesses analyze trends and make decisions based on historical sales data, comparing activity across multiple years.
- **Promotions and Marketing**: Insights from this analysis can be used to design targeted marketing campaigns on slower sales days to increase sales.

## Output Example:

| Year | Weekday | TotalOrders |
| --- | --- | --- |
| 2023 | Tuesday | 600 |
| 2023 | Friday | 550 |
| 2023 | Monday | 500 |
| 2022 | Tuesday | 550 |
| 2022 | Friday | 500 |

The results are grouped by both **year** and **weekday** and ordered by **year** in descending order, followed by the number of orders, showing the most active sales days first.

In [31]:
USE WideWorldImporters;
GO

SELECT 
    YEAR(o.OrderDate) AS Year,                  
    DATENAME(WEEKDAY, o.OrderDate) AS Weekday,  
    COUNT(o.OrderID) AS TotalOrders             
FROM sales.orders AS o
GROUP BY  YEAR(o.OrderDate),                          
          DATENAME(WEEKDAY, o.OrderDate),             
          DATEPART(WEEKDAY, o.OrderDate)             
ORDER BY 
    Year DESC,                                  
    TotalOrders DESC;                           




Commands completed successfully.

(24 rows affected)

Total execution time: 00:00:00.100

Year Weekday TotalOrders 2016 Friday 1848 2016 Thursday 1814 2016 Wednesday 1756 2016 Tuesday 1679 2016 Monday 1676 2016 Saturday 844 2015 Tuesday 4566 2015 Thursday 4372 2015 Wednesday 4167 2015 Monday 4137 2015 Friday 4066 2015 Saturday 2021 2014 Wednesday 4047 2014 Friday 3972 2014 Monday 3836 2014 Thursday 3765 2014 Tuesday 3675 2014 Saturday 1904 2013 Tuesday 3817 2013 Monday 3574 2013 Wednesday 3472 2013 Thursday 3470 2013 Friday 3322 2013 Saturday 1795

# Proposition 5:

This query retrieves the **top 5 best-selling products** in terms of the total quantity sold. It calculates the total units sold for each stock item and orders them from the highest to lowest sales volume. The query provides insights into the top-performing products in terms of sales.

## Key Insights:

- **Total quantity sold** for each product, allowing the business to identify which products have the highest demand.
- The **top 5 best-selling products** are displayed, enabling businesses to focus on their most popular items.
- **Product popularity** can be analyzed to help with inventory management, sales strategy, and marketing efforts.

## Business Use Cases:

- **Inventory Management**: Identifying the top-selling items helps ensure that the business keeps sufficient stock of high-demand products.
- **Sales Performance**: Understanding which products are the most popular helps guide marketing efforts, pricing strategies, and promotional campaigns.
- **Demand Forecasting**: By recognizing which products consistently sell well, businesses can better forecast future demand and optimize their supply chain.

The query ranks the products based on the total quantity sold, showing only the **top 5 products** with the highest sales volume.

In [32]:

USE WideWorldImporters;
GO

SELECT top (5)
    si.StockItemName, 
    SUM(ol.Quantity) AS TotalQuantitySold
FROM sales.orderlines AS ol
JOIN warehouse.StockItems AS si 
    ON ol.StockItemID = si.StockItemID
GROUP BY 
    si.StockItemName
ORDER BY 
    TotalQuantitySold DESC;


Commands completed successfully.

(5 rows affected)

Total execution time: 00:00:00.014

StockItemName TotalQuantitySold Black and orange fragile despatch tape 48mmx75m 207324 Black and orange fragile despatch tape 48mmx100m 193680 Clear packaging tape 48mmx75m 158626 3 kg Courier post bag (White) 300x190x95mm 152375 Shipping carton (Brown) 356x356x279mm 152125

# Proposition 6:

This query retrieves a list of **stock items** that are associated with the color **red** and whose names contain the word **"toy"**. It filters stock items based on both color and product type, specifically targeting toy-related items that are red.

## Key Insights:

- **Stock item names** that include "toy," which allows for a focused search on specific products of interest (e.g., toys).
- The **quantity per outer** of each stock item, which helps in understanding how these toy items are packaged.
- The **color name** is used to filter products that are specifically red, helping to identify color-based products that meet the criteria.

## Business Use Cases:

- **Product Categorization**: Useful for identifying and managing toy products that are of a specific color, such as red, allowing for efficient inventory categorization.
- **Inventory Planning**: Helps warehouse managers plan and manage toy stock, especially for red-colored toy products that might be part of specific promotions or campaigns.
- **Marketing and Sales**: The results can be used for targeted marketing campaigns focused on red-colored toy products, which may appeal to certain customer segments.

The query is filtered to return only stock items that are both **red** in color and include "toy" in the product name, ensuring precise product matching.

In [33]:
USE WideWorldImporters;
GO

select si.StockItemName, si.QuantityPerOuter, ColorName

from warehouse.StockItems as si

JOIN Warehouse.Colors as cl ON cl.ColorID = si.ColorID

where cl.ColorName IN ('red')
AND si.StockItemName LIKE '%toy%'

Commands completed successfully.

(4 rows affected)

Total execution time: 00:00:00.024

StockItemName QuantityPerOuter ColorName RC toy sedan car with remote control (Red) 1/50 scale 1 Red RC vintage American toy coupe with remote control (Red) 1/50 scale 1 Red Ride on toy sedan car (Red) 1/12 scale 1 Red Ride on vintage American toy coupe (Red) 1/12 scale 1 Red

# Proposition 7:

This query provides a list of **customers** who have placed **fewer than 100 orders**. It uses a **Common Table Expression (CTE)** to first calculate the total number of orders per customer, then filters out those customers whose total orders are below 100. The results are ordered by the total number of orders in ascending order, highlighting less frequent customers.

## Key Insights:

- The **total number of orders** placed by each customer, allowing for insights into customer engagement.
- The query focuses on **low-activity customers**, helping businesses understand which customers have made fewer purchases.
- This analysis can reveal **customer segments** that may need re-engagement strategies to increase order frequency.

## Business Use Cases:

- **Customer Retention**: Identifying customers with fewer orders can help the business implement targeted marketing campaigns to encourage more frequent purchases.
- **Sales Analysis**: Understanding the order patterns of low-order customers helps the sales team focus on increasing engagement with these customers.
- **Loyalty Programs**: This data can be used to incentivize customers who are less active by offering promotions or special offers to boost order frequency.

The query retrieves customers who have placed **fewer than 100 orders** and orders them by the **total number of orders** in ascending order, helping the business identify and prioritize low-engagement customers.

In [34]:
USE WideWorldImporters;
GO

WITH CustomerIdCount AS (
    SELECT c.CustomerID,
           c.CustomerName,
           COUNT(OrderID) as [Total Orders]
    FROM sales.Orders as o 
    JOIN sales.Customers as c 
        ON o.CustomerID = c.CustomerID
    GROUP BY
        c.CustomerID,
        c.CustomerName
)

SELECT CustomerID,
       CustomerName,
       [Total Orders]
FROM CustomerIdCount
WHERE [Total Orders] < 100
ORDER BY [Total Orders]


Commands completed successfully.

(96 rows affected)

Total execution time: 00:00:00.025

CustomerID CustomerName Total Orders 1060 Anand Mudaliyar 4 1059 Jibek Juniskyzy 8 1061 Agrita Abele 9 1056 Kalyani Benjaree 13 1058 Jaroslav Fisar 14 1057 Ganesh Majumdar 19 1048 Abhra Ganguly 21 1050 Amrita Ganguly 21 1051 Sylvie Laramee 21 1053 Luis Saucedo 26 1054 Emma Salpa 28 1047 Ivana Hadrabova 28 1055 Adriana Pena 29 1049 Amet Shergill 30 1052 Ian Olofsson 31 1046 Christian Couet 39 1044 Hanita Nookala 43 1036 Erik Malk 50 1038 Damodara Trivedi 51 1039 Bhaamini Palagummi 52 1041 Tomo Vidovic 52 1042 Nguyen Banh 52 1035 Manjunatha Karnik 53 1043 Raj Verma 54 1045 Matteo Cattaneo 57 1040 Damodar Shenoy 59 1030 Chompoo Atitarn 60 1034 Aishwarya Dantuluri 63 1026 Daniella Cavalcante 70 1033 Cuneyt Arslan 72 1028 Emma Van Zant 72 1037 David Jaramillo 73 1032 Som Mukherjee 74 577 Wingtip Toys (Cherryplain, NY) 75 1029 Veronika Necesana 76 1027 Serdar ozden 78 1019 Dhanishta Pullela 78 869 Abel Tatarescu 78 1015 Isa Hulsegge 81 1031 Dipti Shah 81 1020 Hai Banh 82 1012 Urve Kasesalu 82 1021 Fabrice Cloutier 84 1025 Tuulikki Linna 84 1023 Farzana Habibi 86 491 Wingtip Toys (Morton Grove, IL) 87 111 Tailspin Toys (Oriole Beach, FL) 88 1017 David Novacek 88 999 Kristi Part 89 13 Tailspin Toys (Stonefort, IL) 89 39 Tailspin Toys (Diablock, KY) 90 422 Wingtip Toys (Naches, WA) 90 539 Wingtip Toys (Nuangola, PA) 90 818 Liia Orav 91 73 Tailspin Toys (Cokato, MN) 91 95 Tailspin Toys (Lavon, TX) 91 1003 Hue Ton 91 1013 Annette Hetu 91 1022 Nadir Seddigh 91 917 Pari Hosseini 92 922 Sumati Chatterjee 93 865 Bijoya Thakur 93 956 Nhung Thach 93 814 Johanna Hoornstra 93 66 Tailspin Toys (Madaket, MA) 93 490 Wingtip Toys (Beals, ME) 94 511 Wingtip Toys (North Manitou, MI) 94 570 Wingtip Toys (Flomaton, AL) 94 426 Wingtip Toys (Willow Valley, AZ) 94 196 Tailspin Toys (Howells, NE) 94 887 Baran Jonsson 94 904 Olafs Rozitis 95 881 Magdalena Michnova 95 197 Tailspin Toys (Magalia, CA) 95 109 Tailspin Toys (South Laguna, CA) 95 91 Tailspin Toys (Alstead, NH) 95 22 Tailspin Toys (Ward Ridge, FL) 95 533 Wingtip Toys (Nichols Hills, OK) 95 573 Wingtip Toys (Marin City, CA) 96 812 Celica Barajas 96 126 Tailspin Toys (Francis Mills, NJ) 96 477 Wingtip Toys (Glancy, MS) 96 878 Miriam House 96 914 Chandana Shasthri 96 945 Hoc Tran 96 53 Tailspin Toys (Nanafalia, AL) 97 62 Tailspin Toys (Great Neck Estates, NY) 97 839 Juan Morse 97 547 Wingtip Toys (Chaseley, ND) 97 556 Wingtip Toys (Morita, TX) 99 567 Wingtip Toys (Jeromesville, OH) 99 822 Suparna Bhattacharya 99 482 Wingtip Toys (Caton, NY) 99 470 Wingtip Toys (Universal City, TX) 99 883 Radha Barua 99 1024 Arka Chatterjee 99

# Proposition 8:

This query identifies the **top 3 salespersons** based on the **total revenue** they have generated. It uses a **Common Table Expression (CTE)** to calculate the total revenue earned by each salesperson by summing up the revenue from all the orders they handled. The final result shows the top performers in descending order of total revenue.

## Key Insights:

- **Total revenue per salesperson**, which allows for an assessment of individual sales performance.
- **Top 3 salespersons** based on the highest revenue contribution, helping the business focus on high-performing sales staff.
- The **salesperson’s name** is displayed for easier identification, linking revenue directly to team members.

## Business Use Cases:

- **Performance Management**: The query helps management identify the top-performing salespeople who are contributing the most to the company’s revenue. This information can be used for reward programs or performance reviews.
- **Incentive Programs**: The top performers can be incentivized through bonuses or other reward programs based on their high sales contributions.
- **Team Productivity Analysis**: This data can help evaluate the overall productivity of the sales team and determine who is driving the most revenue.

The query returns the **top 3 salespeople** based on their **total revenue**, allowing the business to recognize and reward high-performing employees.

In [35]:
USE WideWorldImporters;
GO

WITH SalespersonRevenue AS (
    SELECT o.SalespersonPersonID, 
           SUM(ol.Quantity * ol.UnitPrice) AS TotalRevenue
    FROM sales.orders AS o
    JOIN sales.orderlines AS ol
        ON o.OrderID = ol.OrderID
    GROUP BY o.SalespersonPersonID
)
SELECT TOP (3)
    p.PersonID, 
    p.FullName AS SalespersonName, 
    sr.TotalRevenue
FROM Application.People AS p
JOIN SalespersonRevenue AS sr 
    ON p.personID = sr.SalespersonPersonID
ORDER BY sr.TotalRevenue DESC


Commands completed successfully.

(3 rows affected)

Total execution time: 00:00:00.043

PersonID SalespersonName TotalRevenue 16 Archer Lamble 18551146.95 2 Kayla Woodcock 18107095.00 3 Hudson Onslow 17815605.10

# Proposition 9:

This query identifies **stock items** that have less than **50 units** on hand. It uses a **Common Table Expression (CTE)** to aggregate the available inventory for each stock item. The final result lists items that are running low on stock, helping with inventory management and restocking decisions.

## Key Insights:

- **Stock item availability**: The query retrieves stock items that have **less than 50 units** available in the warehouse.
- **Product identification**: Each stock item is listed with its name and current quantity on hand, allowing for easy identification of low-stock items.
- **Restocking priorities**: The results are ordered by the quantity on hand, with the lowest stock items appearing first, helping warehouse managers prioritize replenishment.

## Business Use Cases:

- **Inventory Management**: This query helps warehouse managers identify which items need to be restocked based on low inventory levels.
- **Stock Optimization**: By monitoring stock levels and identifying items that are running low, businesses can prevent stockouts and ensure that high-demand products remain available.
- **Demand Forecasting**: Understanding which products are close to being out of stock can help with better planning for supply chain and inventory control.

The query returns all stock items with **less than 50 units** available and orders them by **QuantityOnHand**, making it easy to identify and act on low-stock items.

In [36]:
USE WideWorldImporters;
GO
WITH Inventory AS (
    SELECT 
        si.StockItemID, 
        si.QuantityOnHand
    FROM sales.orderlines AS ol
    JOIN warehouse.StockItemHoldings AS si
        ON ol.StockItemID = si.StockItemID
    JOIN sales.Orders as o 
        ON o.OrderID = ol.OrderID
    GROUP BY si.StockItemID,
    si.QuantityOnHand
)
SELECT 
    si.StockItemID, 
    si.StockItemName, 
    iv.QuantityOnHand
FROM warehouse.StockItems AS si
JOIN Inventory AS iv 
    ON si.StockItemID = iv.StockItemID
WHERE iv.QuantityOnHand < 50
ORDER BY QuantityOnHand ;




Commands completed successfully.

(8 rows affected)

Total execution time: 00:00:00.010

StockItemID StockItemName QuantityOnHand 86 "The Gu" red shirt XML tag t-shirt (White) 5XL 3 78 "The Gu" red shirt XML tag t-shirt (White) XS 16 80 "The Gu" red shirt XML tag t-shirt (White) M 20 204 Tape dispenser (Red) 24 98 "The Gu" red shirt XML tag t-shirt (Black) 4XL 25 77 "The Gu" red shirt XML tag t-shirt (White) XXS 27 184 Shipping carton (Brown) 305x305x305mm 38 95 "The Gu" red shirt XML tag t-shirt (Black) XL 48

# Proposition 10:

This query identifies **orders that were delayed by more than 7 days** past their expected delivery date. It uses a **Common Table Expression (CTE)** to retrieve delayed orders, then filters the results to show only those where the actual picking completion date occurred more than a week after the expected delivery date. This allows for analysis of significantly delayed orders.

## Key Insights:

- **Delayed order details**: The query retrieves the order ID, customer name, order date, expected delivery date, and the actual picking completion date for orders that were delayed by more than 7 days.
- **Customer impact**: By showing the customer name associated with the delayed order, the query helps identify customers who experienced significant delays.
- **Operational efficiency**: The results focus on delays that exceeded 7 days, which can highlight areas where delivery or order processing needs improvement.

## Business Use Cases:

- **Delay Analysis**: This query helps businesses analyze operational inefficiencies by identifying orders that were delayed by more than a week, which may indicate bottlenecks in the fulfillment process.
- **Customer Satisfaction**: Delays over 7 days can affect customer satisfaction, and identifying these delayed orders can help customer service teams proactively address potential issues.
- **Logistics Improvement**: The business can use this information to improve logistics, streamline processes, and reduce long delays in order fulfillment.

The query focuses on orders delayed by **more than 7 days** and orders the results by the most recent **picking completion dates**.

In [37]:
USE WideWorldImporters;
GO

WITH DelayedOrders AS (
    SELECT 
        o.OrderID, 
        o.CustomerID, 
        o.OrderDate, 
        o.ExpectedDeliveryDate, 
        CONVERT(DATE, o.PickingCompletedWhen) AS PickingCompletedDate
    FROM sales.orders AS o
    WHERE o.PickingCompletedWhen > o.ExpectedDeliveryDate
)
SELECT 
    o.OrderID, 
    c.CustomerName, 
    o.OrderDate, 
    o.ExpectedDeliveryDate, 
    o.PickingCompletedDate
FROM DelayedOrders AS o
JOIN sales.customers AS c 
    ON o.CustomerID = c.CustomerID
WHERE DATEDIFF(DAY, o.ExpectedDeliveryDate, o.PickingCompletedDate )> 7
ORDER BY o.PickingCompletedDate DESC;


Commands completed successfully.

(2036 rows affected)

Total execution time: 00:00:00.050

OrderID CustomerName OrderDate ExpectedDeliveryDate PickingCompletedDate 47279 Wingtip Toys (Munich, ND) 2015-04-16 2015-04-17 2016-05-30 47354 Tailspin Toys (Roe Park, NY) 2015-04-18 2015-04-20 2016-05-30 50516 Tailspin Toys (San Acacia, NM) 2015-06-04 2015-06-05 2016-05-30 61152 Veronika Necesana 2015-11-13 2015-11-16 2016-05-30 64503 Narendra Tickoo 2016-01-07 2016-01-08 2016-05-30 63589 Tailspin Toys (Furley, KS) 2015-12-25 2015-12-28 2016-05-27 60177 Pinja Jantunen 2015-10-31 2015-11-02 2016-05-27 60132 Marija Justina Pasek 2015-10-30 2015-11-02 2016-05-27 70554 David Novacek 2016-04-15 2016-04-18 2016-05-26 47075 Tailspin Toys (Goffstown, NH) 2015-04-14 2015-04-15 2016-05-26 71901 Wingtip Toys (Asher, OK) 2016-05-05 2016-05-06 2016-05-25 71997 Tailspin Toys (Eastchester, NY) 2016-05-06 2016-05-09 2016-05-25 72517 Hanita Nookala 2016-05-14 2016-05-16 2016-05-25 48832 Nada Jovanovic 2015-05-11 2015-05-12 2016-05-24 46729 Tailspin Toys (Placer, OR) 2015-04-09 2015-04-10 2016-05-23 44874 Tailspin Toys (Peeples Valley, AZ) 2015-03-11 2015-03-12 2016-05-23 36437 Tailspin Toys (Astor Park, FL) 2014-10-21 2014-10-22 2016-05-23 70458 Aive Petrov 2016-04-14 2016-04-15 2016-05-23 63278 Dhanishta Mokkapati 2015-12-21 2015-12-22 2016-05-23 60799 Tailspin Toys (Panaca, NV) 2015-11-10 2015-11-11 2016-05-23 17445 Wingtip Toys (Paw Paw Lake, MI) 2013-11-21 2013-11-22 2016-05-23 59743 Bhagavaan Atluri 2015-10-23 2015-10-26 2016-05-20 60176 Tailspin Toys (Avenal, CA) 2015-10-31 2015-11-02 2016-05-20 64374 Tailspin Toys (Muir, MI) 2016-01-06 2016-01-07 2016-05-20 50972 Tailspin Toys (Orrtanna, PA) 2015-06-10 2015-06-11 2016-05-19 46471 Wingtip Toys (Obion, TN) 2015-04-06 2015-04-07 2016-05-19 63965 Bhaamini Palagummi 2015-12-31 2016-01-01 2016-05-18 72128 Wingtip Toys (Pikeview, CO) 2016-05-09 2016-05-10 2016-05-18 62839 Wingtip Toys (Pawlet, VT) 2015-12-12 2015-12-14 2016-05-17 62351 Wingtip Toys (Berville, MI) 2015-12-04 2015-12-07 2016-05-16 59386 Wingtip Toys (Trout Run, PA) 2015-10-19 2015-10-20 2016-05-16 70551 Be Trang 2016-04-15 2016-04-18 2016-05-16 48789 Emil Roman 2015-05-09 2015-05-11 2016-05-16 47425 Anca Gogean 2015-04-20 2015-04-21 2016-05-16 49289 Daniel Martensson 2015-05-18 2015-05-19 2016-05-16 70247 Wingtip Toys (Dacono, CO) 2016-04-11 2016-04-12 2016-05-13 71509 Bahaar Asef zade 2016-04-29 2016-05-02 2016-05-13 63520 Amrita Ganguly 2015-12-24 2015-12-25 2016-05-12 50323 Tailspin Toys (East Portal, CO) 2015-06-02 2015-06-03 2016-05-12 47176 Tailspin Toys (Heilwood, PA) 2015-04-15 2015-04-16 2016-05-12 50102 Wingtip Toys (Lake Davis, CA) 2015-05-29 2015-06-01 2016-05-11 58557 Elina Kaleja 2015-10-06 2015-10-07 2016-05-11 71403 Dipti Shah 2016-04-28 2016-04-29 2016-05-11 70558 Anil Sarkar 2016-04-15 2016-04-18 2016-05-09 58889 Tailspin Toys (Airport Drive, MO) 2015-10-12 2015-10-13 2016-05-09 50095 Tailspin Toys (Ikatan, AK) 2015-05-29 2015-06-01 2016-05-09 47705 Wingtip Toys (Cowlington, OK) 2015-04-23 2015-04-24 2016-05-09 37606 Wingtip Toys (Rosa Sánchez, PR) 2014-11-08 2014-11-10 2016-05-09 5645 Wingtip Toys (Munich, ND) 2013-04-19 2013-04-22 2016-05-09 58695 Wingtip Toys (Islip Terrace, NY) 2015-10-08 2015-10-09 2016-05-06 63277 Wingtip Toys (Harkers Island, NC) 2015-12-21 2015-12-22 2016-05-05 46957 Wingtip Toys (Bourbonnais, IL) 2015-04-13 2015-04-14 2016-05-05 49658 Ivan Sepulveda 2015-05-23 2015-05-25 2016-05-04 69155 Tailspin Toys (Statenville, GA) 2016-03-25 2016-03-28 2016-05-04 70102 Wingtip Toys (Oak Point, TX) 2016-04-09 2016-04-11 2016-05-03 49197 Wingtip Toys (Broomtown, AL) 2015-05-16 2015-05-18 2016-05-03 46724 Kumar Naicker 2015-04-09 2015-04-10 2016-05-02 62681 Wingtip Toys (Cavendish, VT) 2015-12-09 2015-12-10 2016-05-02 62109 Tailspin Toys (Lake Stevens, WA) 2015-12-01 2015-12-02 2016-05-02 57777 Tailspin Toys (Gasport, NY) 2015-09-23 2015-09-24 2016-05-02 24641 Wingtip Toys (Delray, WV) 2014-04-04 2014-04-07 2016-05-02 62674 Tailspin Toys (Bratenahl, OH) 2015-12-09 2015-12-10 2016-04-28 48920 Chompo